{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "private_outputs": true,
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "[<img src=\"https://drive.google.com/uc?export=view&id=1D6PD5jhrfNa_D7yfn7ha7OjJslVPy8qg\" width=\"500\" align=\"top\" />](https://drive.google.com/uc?export=view&id=1D6PD5jhrfNa_D7yfn7ha7OjJslVPy8qg)\n",
        "\n",
        "#**ColabPCR**\n",
        "Polymerase Chain Reaction (PCR) amplification of specific genomic regions within eukaryotic genomes frequently exhibits a diminished success rate, often attributed to the non-specific annealing of primers at multiple genomic loci. To mitigate this challenge, we developed ***ColabPCR***, a program meticulously designed for the optimal selection and evaluation of primers targeting a specified genomic region. ColabPCR refines primer length and melting temperature parameters through the utilization of Primer3 software, and it quantifies the number of potential binding regions within the target genome via blastn analysis. Furthermore, the program facilitates the integration of restriction enzyme recognition sites at the 5' termini of primers and provides a mechanism to confirm the absence of such sites within the intended amplification region.\n",
        "\n",
        "\n",
        "###**Input**:  \n",
        "* ***Target Genome***:\n",
        "  * For _Dacus carota_, DCARv2 and DH1v3 genome accessions can be selected from the drop down menu\n",
        "  * Genomes can be uploaded to Google Colab\n",
        "  * Genomes can be downloaded from NCBI Assembly by accession number\n",
        "  * A local file can be specified\n",
        "\n",
        "* Primers are designed using primer3 defaults\n",
        "\n",
        "* Restriction enzymes can be specifed. The selected RE will be added to the 5'end of primers and if they are within the target region available RE will be displayed\n",
        "\n",
        "* positions of the region to amplify: chromosome index `int`, gene start `int`, number of bases from the gene start or end\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "vaKleAjcxToF"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#***Install software and download the selected genome***\n"
      ],
      "metadata": {
        "id": "c6WGJf8tVekw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title ##Install software and load functions\n",
        "import os,sys,json,glob,requests,shutil\n",
        "\n",
        "#@markdown ###Install programs\n",
        "if shutil.which('blastn') is None:\n",
        "  print('Installing BLAST...')\n",
        "  !curl -l ftp://ftp.ncbi.nlm.nih.gov/blast/executables/LATEST/ > ncbi_blast_files.txt 2> /dev/null\n",
        "  blast_file_name = ''\n",
        "  with open('ncbi_blast_files.txt','r') as f:\n",
        "    for line in f:\n",
        "      if ('+-x64-linux.tar.gz' in line) and (not '.md5' in line):\n",
        "        blast_file_name = line.strip()\n",
        "  if len(blast_file_name) > 0:\n",
        "    url = 'https://ftp.ncbi.nlm.nih.gov/blast/executables/LATEST/'+blast_file_name\n",
        "    if not os.path.exists(blast_file_name):\n",
        "      !wget {url} 2> /dev/null 1> logs.txt\n",
        "      !tar -xzvf {blast_file_name} 2> /dev/null 1> logs.txt\n",
        "      blast_folder = blast_file_name.split('+')[0]+'+/bin'\n",
        "    else:\n",
        "      blast_folder = blast_file_name.split('+')[0]+'+/bin'\n",
        "      print('Blast already installed...')\n",
        "  else:\n",
        "    print('Error downloading Blast')\n",
        "  os.environ['PATH'] += f':/content/{blast_folder}'\n",
        "else:\n",
        "  print('Blast already installed...')\n",
        "\n",
        "if shutil.which('primer3_core') is None:\n",
        "  print('Installing Primer3')\n",
        "  !sudo apt-get install -y build-essential g++ cmake git-all -qq 2> /dev/null 1>> logs.txt\n",
        "  !git clone https://github.com/primer3-org/primer3.git primer3\n",
        "  %cd primer3/src\n",
        "  !make all 2> /dev/null 1>> logs.txt\n",
        "  #!make test 2> /dev/null 1>> logs.txt\n",
        "  os.environ['PATH'] += f':/content/primer3/src'\n",
        "  %cd /content\n",
        "else:\n",
        "  print('Primer3 already installed...')\n",
        "\n",
        "if not 'Bio' in sys.modules:\n",
        "  print('Installing Biopython')\n",
        "  !pip install Bio 2> /dev/null 1> logs.txt\n",
        "else:\n",
        "  print('Biopython already installed...')\n",
        "\n",
        "from Bio import Restriction\n",
        "from Bio import SeqIO\n",
        "\n",
        "if not os.path.exists('./datasets'):\n",
        "  print('Installing NCBI Datasets')\n",
        "  !wget https://ftp.ncbi.nlm.nih.gov/pub/datasets/command-line/v2/linux-amd64/datasets 2> /dev/null 1>> logs.txt\n",
        "  !chmod +x datasets\n",
        "else:\n",
        "  print('NCBI Datasets already installed...')\n",
        "\n",
        "#@markdown ###Load functions\n",
        "print('Loading functions...')\n",
        "print('Region extraction function...')\n",
        "\n",
        "def extractRegion(mode,cIndex,Product_Size, start=0,end=0,gpstart = 0,gpend = 0,ntreg=10000, locus_tag='', flexibility=200, fixedNear = False, fixedFar=False):\n",
        "  strand = 1\n",
        "  found = False\n",
        "  for r in SeqIO.parse(gfile,'genbank'):\n",
        "    if r.id == chromosomes[cIndex][0]:\n",
        "      if mode == 'region':\n",
        "        start = start-1\n",
        "        end = end\n",
        "        Product_Size = end-start\n",
        "        print(f'Target location: Start:{start+1} - End:{end}\\n')\n",
        "      elif mode == 'gene':\n",
        "        for f in r.features:\n",
        "          if f.type == 'gene':\n",
        "            if ('locus_tag' in f.qualifiers):\n",
        "              if f.qualifiers[\"locus_tag\"][0] == locus_tag:\n",
        "                  lstart = f.location.start\n",
        "                  lend = f.location.end\n",
        "                  strand = f.location.strand\n",
        "                  if gpend == 0:\n",
        "                    gpend = abs(lend-lstart)\n",
        "                  if strand == 1:\n",
        "                    start = lstart+gpstart\n",
        "                    end = lstart+gpend\n",
        "                  else:\n",
        "                    end = lend-gpstart\n",
        "                    start = lend-gpend\n",
        "                  Product_Size = abs(lend-lstart)\n",
        "                  print(f'Target location: Start:{start+1} - End:{end}\\nLocus tag location: Start:{lstart+1} - End:{lend} strand {strand}')\n",
        "      else:\n",
        "        for f in r.features:\n",
        "          if f.type == 'CDS':\n",
        "            if ('locus_tag' in f.qualifiers):\n",
        "              if f.qualifiers[\"locus_tag\"][0] == locus_tag:\n",
        "                if mode == 'promoter':\n",
        "                  found = True\n",
        "                  lstart = f.location.start\n",
        "                  lend = f.location.end\n",
        "                  strand = f.location.strand\n",
        "                  if strand == 1:\n",
        "                    start = lstart-Product_Size\n",
        "                    end = lstart\n",
        "                  else:\n",
        "                    start = lend\n",
        "                    end = lend+Product_Size\n",
        "                  print(f'Target location: Start:{start+1} - End:{end}\\nLocus tag location: Start:{lstart+1} - End:{lend} strand {strand}')\n",
        "                elif mode == 'terminator':\n",
        "                  found = True\n",
        "                  lstart = f.location.start\n",
        "                  lend = f.location.end\n",
        "                  strand = f.location.strand\n",
        "                  if strand == 1:\n",
        "                    start = lend\n",
        "                    end = lend+Product_Size\n",
        "                  else:\n",
        "                    start = lstart-Product_Size\n",
        "                    end = lstart\n",
        "                  print(f'Target location: Start:{start+1} - End: {end}\\nLocus tag location: Start:{lstart+1} - End:{lend} strand {strand}')\n",
        "        ###\n",
        "        if not found:\n",
        "          raise Exception(f'{locus_tag} not found.')\n",
        "      ###\n",
        "      targetSeq = r[start:end]\n",
        "      #Checks for overlapping CDSs\n",
        "      print(f'The following features are in the +- {ntreg} nucleotides range:')\n",
        "      ltags = []\n",
        "      for f in r[start-ntreg:end+ntreg].features:\n",
        "        if ('locus_tag' in f.qualifiers):\n",
        "          #if f.type == 'CDS':\n",
        "          if not f.qualifiers[\"locus_tag\"][0] == locus_tag:\n",
        "            if not f.qualifiers[\"locus_tag\"][0] in ltags:\n",
        "              ltags.append(f.qualifiers[\"locus_tag\"][0])\n",
        "              print(f'--> {f.qualifiers[\"locus_tag\"][0]}-{start+1-ntreg+f.location.start}-{start-ntreg+f.location.end}')\n",
        "              feat_len = abs(f.location.end - f.location.start)\n",
        "              feat_start = min(f.location.end,f.location.start)\n",
        "              feat_end = max(f.location.end,f.location.start)\n",
        "              if (feat_start in range(ntreg,ntreg+Product_Size)) or (feat_end -1 in range(ntreg,ntreg+Product_Size)):\n",
        "                print(f'  * This feature overlaps the selected promoter region... Start: {start+1-ntreg+feat_start} - End:{start-ntreg+feat_end} Range: {start+1}-{start+Product_Size}')\n",
        "      ###\n",
        "      if flexibility > Product_Size/2:\n",
        "        flexibility = int(Product_Size/3)\n",
        "        print(f'The program will allow to move the region bounds {flexibility} nucleotides on each side')\n",
        "      ###default return values\n",
        "      inStart = start\n",
        "      inEnd = end\n",
        "      FORCE_RIGHT_START = False\n",
        "      FORCE_LEFT_START = False\n",
        "      ###\n",
        "      if  mode == 'region':\n",
        "        if fixedNear and not fixedFar and (strand == 1):\n",
        "          print(f'Fixing the primer on the start codon side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inEnd = end - start - 1 - flexibility\n",
        "          inStart = 0\n",
        "          FORCE_LEFT_START = True\n",
        "        elif fixedFar and not fixedNear and (strand == 1):\n",
        "          print(f'Fixing the primer on the distant side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inStart = flexibility\n",
        "          inEnd = end - start - 1\n",
        "          FORCE_RIGHT_START = True\n",
        "        elif fixedFar and fixedNear:\n",
        "          print(f'Fixing the primer on the both sides. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inStart = 0\n",
        "          inEnd = end - start - 1\n",
        "          FORCE_RIGHT_START = True\n",
        "          FORCE_LEFT_START = True\n",
        "        else:\n",
        "          print(f'Both target ends set to flexible ({flexibility}). Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inStart = flexibility\n",
        "          inEnd = end - start - 1 - flexibility\n",
        "      elif  mode == 'gene':\n",
        "        if fixedNear and not fixedFar and (strand == 1):\n",
        "          print(f'Fixing the primer on the start codon side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inEnd = end - start - 1 - flexibility\n",
        "          inStart = 0\n",
        "          FORCE_LEFT_START = True\n",
        "        elif fixedNear and not fixedFar and (strand == -1):\n",
        "          print(f'Fixing the primer on the start codon side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inEnd = end - start - 1\n",
        "          inStart = flexibility\n",
        "          FORCE_RIGHT_START = True\n",
        "        elif fixedFar and not fixedNear and (strand == 1):\n",
        "          print(f'Fixing the primer on the distant side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inStart = flexibility\n",
        "          inEnd = end - start - 1\n",
        "          FORCE_RIGHT_START = True\n",
        "        elif fixedFar and not fixedNear and (strand == -1):\n",
        "          print(f'Fixing the primer on the distant side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inEnd = end - start - 1 - flexibility\n",
        "          inStart = 0\n",
        "          FORCE_LEFT_START = True\n",
        "        elif fixedFar and fixedNear:\n",
        "          print(f'Fixing the primer on the both sides. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inStart = 0\n",
        "          inEnd = end - start - 1\n",
        "          FORCE_RIGHT_START = True\n",
        "          FORCE_LEFT_START = True\n",
        "        else:\n",
        "          print(f'Both target ends set to flexible ({flexibility}). Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inStart = flexibility\n",
        "          inEnd = end - start - 1 - flexibility\n",
        "      ###\n",
        "      elif  mode == 'promoter':\n",
        "        if fixedNear and not fixedFar and (strand == 1):\n",
        "          print(f'Fixing the primer on the start codon side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inEnd = end - start - 1\n",
        "          inStart = flexibility\n",
        "          FORCE_RIGHT_START = True\n",
        "        elif fixedNear and not fixedFar and (strand == -1):\n",
        "          print(f'Fixing the primer on the start codon side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inEnd = end - start - 1 - flexibility\n",
        "          inStart = 0\n",
        "          FORCE_LEFT_START = True\n",
        "        elif fixedFar and not fixedNear and (strand == 1):\n",
        "          print(f'Fixing the primer on the distant side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inStart = 0\n",
        "          inEnd = end - start - 1 - flexibility\n",
        "          FORCE_LEFT_START = True\n",
        "        elif fixedFar and not fixedNear and (strand == -1):\n",
        "          print(f'Fixing the primer on the distant side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inEnd = end - start - 1\n",
        "          inStart = flexibility\n",
        "          FORCE_RIGHT_START = True\n",
        "        elif fixedFar and fixedNear:\n",
        "          print(f'Fixing the primer on the both sides. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inStart = 0\n",
        "          inEnd = end - start - 1\n",
        "          FORCE_RIGHT_START = True\n",
        "          FORCE_LEFT_START = True\n",
        "        else:\n",
        "          print(f'Both target ends set to flexible ({flexibility}). Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inStart = flexibility\n",
        "          inEnd = end - start - 1 - flexibility\n",
        "      ###\n",
        "\n",
        "      elif  mode == 'terminator':\n",
        "        if fixedNear and not fixedFar and (strand == 1):\n",
        "          print(f'Fixing the primer on the start codon side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inEnd = end - start - 1 - flexibility\n",
        "          inStart = 0\n",
        "          FORCE_LEFT_START = True\n",
        "        elif fixedNear and not fixedFar and (strand == -1):\n",
        "          print(f'Fixing the primer on the start codon side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inEnd = end - start - 1 - flexibility\n",
        "          inStart = flexibility\n",
        "          FORCE_RIGHT_START = True\n",
        "        elif fixedFar and not fixedNear and (strand == 1):\n",
        "          print(f'Fixing the primer on the distant side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inStart = flexibility\n",
        "          inEnd = end - start - 1\n",
        "          FORCE_RIGHT_START = True\n",
        "        elif fixedFar and not fixedNear and (strand == -1):\n",
        "          print(f'Fixing the primer on the distant side. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inEnd = end - start - 1 -flexibility\n",
        "          inStart = 0\n",
        "          FORCE_LEFT_START = True\n",
        "        elif fixedFar and fixedNear:\n",
        "          FORCE_RIGHT_START = True\n",
        "          FORCE_LEFT_START = True\n",
        "          print(f'Fixing the primer on the both sides. Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inStart = 0\n",
        "          inEnd = end - start - 1\n",
        "        else:\n",
        "          print(f'Both target ends set to flexible ({flexibility}). Start: {start+1} End: {end} Strand: {strand}')\n",
        "          inStart = flexibility\n",
        "          inEnd = end - start - 1 - flexibility\n",
        "      ###\n",
        "      if (not fixedNear) and (not fixedFar):\n",
        "        prangemin = end - start - 2*flexibility\n",
        "      elif (not fixedNear) or (not fixedFar):\n",
        "        prangemin = end - start - flexibility\n",
        "      else:\n",
        "        prangemin = end - start\n",
        "      prangemax = end - start\n",
        "      targetSeq = r[start:end]\n",
        "      SeqIO.write(targetSeq, 'target.fasta','fasta')\n",
        "      print(f'Target product size (target.fasta): {len(targetSeq)}')\n",
        "      print(f'Flexible Product size range: {prangemin}-{prangemax}')\n",
        "      print(f'{targetSeq.seq}')\n",
        "      return(targetSeq,inStart,inEnd,prangemin,prangemax,FORCE_RIGHT_START,FORCE_LEFT_START)\n",
        "  print(f'Error extracting region.')\n",
        "\n",
        "##virtual PCR functions\n",
        "print('Virtual PCR functions...')\n",
        "\n",
        "import pandas as pd\n",
        "\n",
        "def PCR(q,oligo_5_prima_Fwd,oligo_5_prima_rv,gfile,max_diff_aln_len,remove_oligos):\n",
        "  qids = []\n",
        "  dbname = os.path.splitext(os.path.basename(gfile))[0]\n",
        "  if remove_oligos:\n",
        "    for r in SeqIO.parse(q,'fasta'):\n",
        "      if str(r.seq).startswith(oligo_5_prima_rv):\n",
        "        r.seq = r.seq[len(oligo_5_prima_rv):]\n",
        "      elif str(r.seq).startswith(oligo_5_prima_Fwd):\n",
        "        r.seq = r.seq[len(oligo_5_prima_Fwd):]\n",
        "      qids.append(r)\n",
        "    SeqIO.write(qids,'query.fasta','fasta')\n",
        "    query = 'query.fasta'\n",
        "  else:\n",
        "    query = q\n",
        "    max_diff_aln_len += max(len(oligo_5_prima_rv),len(oligo_5_prima_Fwd))\n",
        "\n",
        "  #Blast\n",
        "  cpu_count = os.cpu_count()\n",
        "  if cpu_count > 1:\n",
        "    if not os.path.exists(f'{dbname}.nsq'):\n",
        "      fnaFile = os.path.splitext(gfile)[0]+'.fna'\n",
        "      SeqIO.convert(gfile,'gb',fnaFile,'fasta')\n",
        "      os.system(f'makeblastdb -in {fnaFile} -dbtype nucl -out {dbname}')\n",
        "    else:\n",
        "      print(f'Using existent blast database: {dbname} ')\n",
        "    os.system(f'blastn -query {query} -db {dbname} -task blastn-short -word_size 4 -evalue 1 -outfmt \"6 qseqid sseqid length qlen qstart qend slen sstart send sstrand qseq sseq evalue pident mismatch qcovs qcovhsp qcovus\" -max_target_seqs 10 -out blast.res -num_threads {cpu_count}')\n",
        "  else:\n",
        "    if not os.path.exists(fnaFile):\n",
        "      fnaFile = os.path.splitext(gfile)[0]+'.fna'\n",
        "      SeqIO.convert(gfile,'gb',fnaFile,'fasta')\n",
        "    os.system(f'blastn -query {query} -subject {fnaFile} -task blastn-short -word_size 4 -evalue 1 -outfmt \"6 qseqid sseqid length qlen qstart qend slen sstart send sstrand qseq sseq evalue pident mismatch qcovs qcovhsp qcovus\" -max_target_seqs 10 -out blast.res')\n",
        "  blast = pd.read_csv('blast.res',sep='\\t',names=['qseqid','sseqid','length','qlen','qstart','qend','slen','sstart','send','sstrand', 'qseq','sseq','evalue','pident','mismatch','qcovs','qcovhsp','qcovus'], index_col=False)\n",
        "  blast = blast[blast['length'] >= blast['qlen']-max_diff_aln_len]\n",
        "  return(blast)\n",
        "\n",
        "def primer_blast(q,oligo_5_prima_Fwd,oligo_5_prima_rv,gfile,primerPair,max_diff_aln_len,remove_oligos):\n",
        "  #correr blast para encontrar hits de primers en genoma\n",
        "  blast = PCR(q,oligo_5_prima_Fwd,oligo_5_prima_rv,gfile,max_diff_aln_len,remove_oligos)\n",
        "  blast.to_csv(f'FilteredPrimerHits_{primerPair}.csv')\n",
        "\n",
        "  #\n",
        "  qseqids = blast['qseqid'].unique()\n",
        "  if len(qseqids) != 2:\n",
        "    sys.exit(f'Error: Either the query has more than two primers, or only one primer was found on the target genome...')\n",
        "  else:\n",
        "    fwd = qseqids[0]\n",
        "    rv = qseqids[1]\n",
        "  #new 7/11/2024 Counts number of hits\n",
        "  fwd_found_no_mismatch = len(blast[(blast['qseqid'] == fwd) & (blast['mismatch'] == 0)])\n",
        "  rv_found_no_mismatch = len(blast[(blast['qseqid'] == rv) & (blast['mismatch'] == 0)])\n",
        "  fwd_found_with_mismatch = len(blast[(blast['qseqid'] == fwd) & (blast['mismatch'] > 0)])\n",
        "  rv_found_with_mismatch = len(blast[(blast['qseqid'] == rv) & (blast['mismatch'] > 0)])\n",
        "  #end_new\n",
        "  #analizar tabla de blast y buscar resultados compatibles con productos de amplificaciÃ³n.\n",
        "  with open('virtualPCR.csv','a+') as f:\n",
        "    #new 7/11/2024_Table headers\n",
        "    #end_new\n",
        "    numProducts = 0\n",
        "    sseqids = blast['sseqid'].unique()\n",
        "    for sid in sseqids:\n",
        "      sblast = blast[blast['sseqid'] == sid]\n",
        "      if len(sblast) >= 2:\n",
        "        sfwdblast = sblast[sblast['qseqid']==fwd]\n",
        "        srvblast = sblast[sblast['qseqid']==rv]\n",
        "        if (len(sfwdblast) > 0 ) and (len(srvblast) > 0):\n",
        "          for _,fwdHit in sfwdblast.iterrows():\n",
        "            if fwdHit['sstrand'] == 'plus':\n",
        "              rvStrand = 'minus'\n",
        "              start = fwdHit['sstart']\n",
        "            elif fwdHit['sstrand'] == 'minus':\n",
        "              rvStrand = 'plus'\n",
        "              end = fwdHit['sstart']\n",
        "            srvblastMinus = srvblast[srvblast['sstrand'] == rvStrand]\n",
        "            if len(srvblastMinus) > 0:\n",
        "              for _,rvHit in srvblastMinus.iterrows():\n",
        "                if rvHit['sstrand'] == 'minus':\n",
        "                  if rvHit['send'] > fwdHit['send']:\n",
        "                    end = rvHit['sstart']\n",
        "                    pcrProducLen = end - start + 1\n",
        "                  else:\n",
        "                    end = rvHit['sstart']\n",
        "                    pcrProducLen = 0\n",
        "                elif rvHit['sstrand'] == 'plus':\n",
        "                  if rvHit['send'] < fwdHit['send']:\n",
        "                    start = rvHit['sstart']\n",
        "                    pcrProducLen = end - start + 1\n",
        "                  else:\n",
        "                    start = rvHit['sstart']\n",
        "                    pcrProducLen = 0\n",
        "                #Primer Fwd,Primer Fwd length,Fwd Aln Length, Fwd Hit QSeq, Fwd Hit SSeq,\n",
        "                #Fwd Percent ID, Fwd Mismatches, Primer Rv, Primer Rv length,\n",
        "                #Rv Aln Length, Rv Hit QSeq, Rv Hit SSeq, Rv Percent ID, Rv Mismatches, Product Length'\n",
        "                #new_updated_7/11/2024 to add fwd and rv hjit counts\n",
        "                f.write(f\"{fwdHit['sseqid']},{start},{end},{fwdHit['qseqid']},{fwdHit['qlen']},{fwd_found_no_mismatch},{fwd_found_with_mismatch},{fwdHit['length']},{fwdHit['qseq']},{fwdHit['sseq']},{fwdHit['pident']},{fwdHit['mismatch']},{rvHit['qseqid']},{rvHit['qlen']},{rv_found_no_mismatch},{rv_found_with_mismatch},{rvHit['length']},{rvHit['qseq']},{rvHit['sseq']},{rvHit['pident']},{rvHit['mismatch']}\")\n",
        "                #end_new\n",
        "                if pcrProducLen == 0:\n",
        "                  f.write(f',Divergent primers\\n')\n",
        "                elif pcrProducLen < 20000:\n",
        "                  f.write(f',{pcrProducLen}\\n')\n",
        "                  numProducts+=1\n",
        "                else:\n",
        "                  f.write(f',{pcrProducLen} Long PCR Product\\n')\n",
        "\n",
        "  print(f'Number of PCR products of length < 20000: {numProducts}.')\n",
        "  return(numProducts)\n",
        "\n",
        "#validate_oligos\n",
        "def validate_oligo(oligo):\n",
        "  if len(oligo) > 0:\n",
        "    for i in oligo:\n",
        "      if i not in 'ATCG':\n",
        "        return(False)\n",
        "    return(True)\n",
        "  else:\n",
        "    return(False)\n"
      ],
      "metadata": {
        "id": "0pnEjCKSxSs_",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title ## Download target genome\n",
        "\n",
        "#@markdown ### Select target genome mode from the dropdown menu. </br>\n",
        "#@markdown ##### To download the target genome from [NCBI Assembly/Dataset genomes](https://www.ncbi.nlm.nih.gov/datasets/genome/) use: NCBI Assembly Accession number.\n",
        " #@markdown * **Accsession number example**: GCA_001625215.2 [*Daucus carota*]\n",
        "#@markdown ##### To download *Daucus carota* genome select the Genome version from the dropdown menu.\n",
        "#@markdown ##### To upload a genome in genebank format, select: upload\n",
        "#@markdown ##### To use a local file, select: local_file\n",
        "#@markdown <br>\n",
        "Genome_Mode = 'NCBI Assembly Accession number' #@param [\"NCBI Assembly Accession number\",\"upload file\",\"local_file\",\"Daucus carota DH1v3\", \"Daucus carota DCARv2\"]\n",
        "#@markdown <br>\n",
        "\n",
        "from Bio import SeqIO\n",
        "import os\n",
        "\n",
        "def downloadGenome(acc,gfile):\n",
        "  if not os.path.exists(gfile):\n",
        "    !./datasets download genome accession {acc} --include gbff --filename {acc}.zip >/dev/null 2>&1\n",
        "    !unzip -o -j {acc}.zip ncbi_dataset/data/{acc}/*.gbff -d ./ >/dev/null 2>&1\n",
        "    os.remove(f'{acc}.zip')\n",
        "    !mv genomic.gbff {gfile} >/dev/null 2>&1\n",
        "    pass\n",
        "  else:\n",
        "    pass\n",
        "\n",
        "\n",
        "\n",
        "#@markdown ##### If you selected **\"NCBI Assembly Accession number\"** please input the accession number in the form.\n",
        "NCBI_genome_Accession = \"\" #@param {type:'string'}\n",
        "#@markdown ##### <br> If you selected **\"Local file\"** please input the path to the genome in genbank format.\n",
        "local_file = \"\" #@param {type:'string'}\n",
        "\n",
        "if Genome_Mode == 'NCBI Assembly Accession number':\n",
        "  #\n",
        "  if len(NCBI_genome_Accession) > 0:\n",
        "    if 'GCA' in NCBI_genome_Accession or 'GCF' in NCBI_genome_Accession:\n",
        "      acc = NCBI_genome_Accession\n",
        "      print(f'Downloading {acc}')\n",
        "      gfile = f'{acc}.gb'\n",
        "      downloadGenome(acc,gfile)\n",
        "    else:\n",
        "      raise Exception(f'{NCBI_genome_Accession} is not a NCBI assembly accession number.')\n",
        "elif Genome_Mode == 'Daucus carota DH1v3':\n",
        "  print('Downloading genome:\\n--> DH1v3: GCA_001625215.2')\n",
        "  acc = 'GCA_001625215.2'\n",
        "  gfile = 'dh3v1.gb'\n",
        "  downloadGenome(acc,gfile)\n",
        "elif Genome_Mode == 'Daucus carota DCARv2':\n",
        "  print('Downloading genome:\\n--> DCARv2: GCA_001625215.1')\n",
        "  gfile = 'dcarv2.gb'\n",
        "  acc = 'GCA_001625215.1'\n",
        "  downloadGenome(acc,gfile)\n",
        "elif Genome_Mode == 'upload file':\n",
        "  from google.colab import files\n",
        "  uploaded = files.upload()\n",
        "  gfile = uploaded.keys()[0]\n",
        "  try:\n",
        "    gfileName = os.path.splitext(gfile)[0]+\".gb\"\n",
        "    SeqIO.convert(gfile,'fasta',gfileName,'gb')\n",
        "    os.remove(gfile)\n",
        "    gfile = gfileName\n",
        "    print(f'Uploaded file {gfile} converted to Genbank format.\\nWarning: Locus_tag region extraction requires an annotated genome.')\n",
        "  except:\n",
        "    print(f'Uploaded file {gfile}')\n",
        "elif Genome_Mode == 'local_file':\n",
        "  if os.path.exists(local_file):\n",
        "    try:\n",
        "      gfileName = os.path.splitext(local_file)[0]+\".gb\"\n",
        "      SeqIO.convert(local_file,'fasta',gfileName,'gb')\n",
        "      os.remove(local_file)\n",
        "      local_file = gfileName\n",
        "      print(f'{local_file} converted to Genbank format.\\nWarning: Locus_tag region extraction requires an annotated genome.')\n",
        "    except:\n",
        "      print(f'Local file loaded: {local_file}')\n",
        "  else:\n",
        "    print(f'Error: {local_file} does not exist.')\n",
        "else:\n",
        "  raise Exception(f'Error: the selected genome could not be downloaded.')\n",
        "\n",
        "#Get the list of chromosomes/scaffolds/contigs\n",
        "chromosomes = {}\n",
        "\n",
        "for i,r in enumerate(SeqIO.parse(gfile,'genbank')):\n",
        "  chromosomes[i] = [r.id,r.description]\n",
        "\n",
        "for i in chromosomes:\n",
        "  print(f'Chromosome index [cindex] {i}: {chromosomes[i][0]} - {chromosomes[i][1]}')\n",
        "\n"
      ],
      "metadata": {
        "id": "qXl5qEyL8lpl",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n"
      ],
      "metadata": {
        "id": "mlZiyas9NiYh"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#**Extract target sequence from the genome**"
      ],
      "metadata": {
        "id": "Yx4shmk-VqT9"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title #Extract the target sequence from the downloaded genome\n",
        "#@markdown ### **Requiered data**:\n",
        "#@markdown - chromosome index form the previous list [0 indexed]\n",
        "#@markdown - start and end location [for region mode]\n",
        "#@markdown - locus_tag [for gene and promoter/terminator mode]\n",
        "#@markdown - start and end positions within a gene [for gene mode]\n",
        "#@markdown - Product size [for promoter/terminator mode]\n",
        "#@markdown - primer region end flexibility\n",
        "\n",
        "#@markdown <br>\n",
        "import sys\n",
        "breake_execution = False\n",
        "flexibility = 0\n",
        "fixedFar = False\n",
        "fixedNear = False\n",
        "\n",
        "#@markdown ###**Select the running mode**\n",
        "mode = 'region' #@param ['promoter','terminator','region','gene']\n",
        "\n",
        "#@markdown #### <br/>**Chromosome index** [Required for all running modes]\n",
        "cIndex = 0 #@param {type:\"integer\"}\n",
        "if cIndex == None or not isinstance(cIndex,int):\n",
        "  breake_execution = True\n",
        "\n",
        "#@markdown #### <br/>**Locus tag** [Only used in the gene and Promoter/terminator modes]\n",
        "locus_tag = ''\n",
        "if mode == 'gene' or mode == 'promoter' or mode == 'terminator':\n",
        "  locus_tag = ''  #@param {type:\"string\"}\n",
        "\n",
        "#@markdown ####<br/> **Start and end genomic coordinates** [Required only in the region mode]\n",
        "start, end = 0,0\n",
        "start_position_within_gene = 0\n",
        "end_position_within_gene = 0\n",
        "\n",
        "\n",
        "if mode == 'region':\n",
        "  start = 5000  #@param {type:\"integer\"}\n",
        "  end = 10000  #@param {type:\"integer\"}\n",
        "  if start == None or end == None or not isinstance(start,int) or not isinstance(end,int):\n",
        "    start, end = 0,0\n",
        "  Product_Size = abs(start-end)\n",
        "  print(f'Target region: {cIndex}-{chromosomes[cIndex][0]}:{start}, {end}')\n",
        "  if Product_Size <= 100:\n",
        "    print(f'The specified Target region is too short. Are the start and end positions correct?')\n",
        "    breake_execution = True\n",
        "  if start > end:\n",
        "    print(f'The start position is greater than the end position. Are the start and end positions correct?')\n",
        "    breake_execution = True\n",
        "elif mode == 'gene':\n",
        "  #@markdown #### <br/>**Extract a subsequence of the gene [start and end nucleotides relative to the gene start]** <br/>For full gene length use: start_position_within_gene=0, end_position_within_gene=0 <br>[Only used in the gene mode]\n",
        "  start_position_within_gene = 0 #@param {type:\"integer\"}\n",
        "  end_position_within_gene = 0 #@param {type:\"integer\"}\n",
        "  Product_Size = 0\n",
        "else:\n",
        "  #@markdown ### <br> **Options for promoter/terminator modes**\n",
        "\n",
        "  #@markdown #### <br> ***Product size***. Number of nucleotide of the promoter or terminator region to extract.<br/> [Only in the promoter and terminator modes]\n",
        "  Product_Size = 500 #@param {type:\"integer\"}\n",
        "  if Product_Size == None or not isinstance(Product_Size,int):\n",
        "    breake_execution = True\n",
        "\n",
        "#@markdown ### <br> **General options**\n",
        "#@markdown ##### ***Flexibility:*** how far away (nts) from the start/end coordinates can the designed primers bind. This option is for both sides unless fixedNear or fixedFar is selected.\n",
        "flexibility = 200 #@param {type:\"integer\"}\n",
        "if flexibility == None or not isinstance(flexibility,int):\n",
        "  breake_execution = True\n",
        "#@markdown FixedNear: the primer near start (promoter mode) or end (terminator mode) codon will be fixed. <br>\n",
        "#@markdown FixedFar: the distant primer will be fixed.\n",
        "fixedFar = False #@param {type:\"boolean\"}\n",
        "fixedNear = False #@param {type:\"boolean\"}\n",
        "\n",
        "if breake_execution:\n",
        "  print('Error extracting target region...')\n",
        "else:\n",
        "  targetSeq,inStart,inEnd,prangemin,prangemax,FORCE_RIGHT_START,FORCE_LEFT_START = extractRegion(mode,cIndex,Product_Size,locus_tag=locus_tag,start=start,end=end, gpstart=start_position_within_gene, gpend=end_position_within_gene,fixedFar=fixedFar,fixedNear=fixedNear)\n"
      ],
      "metadata": {
        "id": "XtIM5JRwBKW2",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title #Use a custom sequence [cut and paste]\n",
        "#@markdown ###**Please input the target sequence [ATCG]**\n",
        "target_sequence_id = '' #@param {type:\"string\"}\n",
        "target_sequence = '' #@param {type:\"string\"}\n",
        "target_sequence = target_sequence.upper()\n",
        "\n",
        "with open('target.fasta','w') as f:\n",
        "  f.write(f'>{target_sequence_id }\\n{target_sequence}\\n')\n",
        "\n",
        "from Bio import SeqIO\n",
        "targetSeq = SeqIO.read('target.fasta','fasta')\n",
        "\n",
        "#@markdown ***Flexibility***: allows the program to design primers that bind a number of specified nucleotides away from the sequence ends. This option is used in both sides unless fixedLeft or fixedRight is selected.\n",
        "\n",
        "flexibility = 200 #@param {type:\"integer\"}\n",
        "#@markdown Select fixedLeft if you want the left primer to be anchored in the 5'-end of the target sequence. Select fixedRight if you want the right primer to be anchored in the 3'-end of the target sequence.\n",
        "fixedLeft = False #@param {type:\"boolean\"}\n",
        "fixedRight = False #@param {type:\"boolean\"}\n",
        "\n",
        "FORCE_RIGHT_START = fixedRight\n",
        "FORCE_LEFT_START = fixedLeft\n",
        "\n",
        "if (not fixedLeft) and (not fixedRight):\n",
        "  prangemin = len(targetSeq.seq) - 2*flexibility\n",
        "elif fixedLeft and (not fixedRight):\n",
        "  prangemin = len(targetSeq.seq) - flexibility\n",
        "  inStart = 0\n",
        "  inEnd = len(targetSeq.seq) - flexibility\n",
        "elif (not fixedLeft) and fixedRight:\n",
        "  prangemin = len(targetSeq.seq) - flexibility\n",
        "  inStart = flexibility\n",
        "  inEnd = len(targetSeq.seq) - 1\n",
        "else:\n",
        "  inStart = flexibility\n",
        "  inEnd = len(targetSeq.seq) - flexibility\n",
        "  prangemin = len(targetSeq.seq)\n",
        "\n",
        "prangemax = len(targetSeq.seq)\n",
        "print(f'Target product size (target.fasta): {len(targetSeq)}')\n",
        "print(f'Flexible Product size range: {prangemin}-{prangemax}')\n",
        "print(f'{targetSeq.seq}')"
      ],
      "metadata": {
        "id": "kEBSPa0OkgqQ",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#***Evaluate the presence of restriction enzyme sites in the target sequence and add custom oligos in the 5'end of primers***"
      ],
      "metadata": {
        "id": "UilctXLZpJZl"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title ##Evaluate the presence of REs sites in the target sequence\n",
        "from Bio import Restriction\n",
        "\n",
        "rb_supp = Restriction.RestrictionBatch(first=[], suppliers=['C','B','E','I','K','J','M',\n",
        "'O','N','Q','S','R','V','Y','X'])\n",
        "\n",
        "cutters = [r for r,s in rb_supp.search(targetSeq.seq).items() if len(s) > 0]\n",
        "noncutters = [r for r,s in rb_supp.search(targetSeq.seq).items() if len(s) == 0]\n",
        "\n",
        "#filter results by list\n",
        "#@markdown ### **Add the Restriction enzyme names as a comma separated list.**\n",
        "ER_list = 'EcoRI,BamHI,HindIII,XbaI' #@param {type:'string'}\n",
        "ER_list = ER_list.replace(' ','')\n",
        "ER_list = ER_list.split(',')\n",
        "\n",
        "print(f'Analyzing sequence for restriction sites... {ER_list}')\n",
        "\n",
        "for e in cutters:\n",
        "  if str(e) in ER_list:\n",
        "    print(f'X  --> The restriction enzyme {e} cuts the target sequence.')\n",
        "\n",
        "for e in noncutters:\n",
        "  if str(e) in ER_list:\n",
        "    print(f'OK --> The restriction enzyme {e} does not cut the target sequence.')\n",
        "\n",
        "print_all_noncutters = False #@param {type:\"boolean\"}\n",
        "if print_all_noncutters or len(ER_list) == 0:\n",
        "  print('\\nAll noncutters RE:')\n",
        "  for i in range(0,len(noncutters),10):\n",
        "    for e in noncutters[i:i+10]:\n",
        "      print(f'{str(e)}  ', end=(''))\n",
        "    print()"
      ],
      "metadata": {
        "cellView": "form",
        "id": "KUrKNHJ4B5Wy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title ##Add 5' Oligonucleotides and restriction enzyme regcognition sites to the *fwd* and *rv* primers\n",
        "#@markdown The forward (fwd) primer binds on the 5' end of the target sequence. <br> The reverse (rv) binds in the 3' end of the targetr sequence.\n",
        "\n",
        "\n",
        "#@markdown 5'- [oligo\\_5'-] [ER] [Primer Sequence] -3'<br>\n",
        "#@markdown Input the oligonucleotides, or Restriction enzyme by name (eg. EcoRI, case matters)\n",
        "\n",
        "\n",
        "oligo_5_prime_Fwd = 'AA' #@param {type:\"string\"}\n",
        "\n",
        "if not validate_oligo(oligo_5_prime_Fwd) and oligo_5_prime_Fwd != '':\n",
        "  print(f'Error: {oligo_5_prime_Fwd} is not a valid oligo sequence.')\n",
        "  oligo_5_prime_Fwd = ''\n",
        "\n",
        "\n",
        "enzyme_fwd = 'EcoRI' #@param {type:'string'}\n",
        "if not enzyme_fwd == '':\n",
        "  if enzyme_fwd in rb_supp:\n",
        "    enzime_site_f = rb_supp.get(enzyme_fwd).site\n",
        "    oligo_5_prime_Fwd2 = oligo_5_prime_Fwd + enzime_site_f\n",
        "    print(f\"Oligo 5'-Fwd: {oligo_5_prime_Fwd} {enzime_site_f} --> primer sequence'\")\n",
        "  elif validate_oligo(enzyme_fwd):\n",
        "    oligo_5_prime_Fwd2 = oligo_5_prime_Fwd + enzyme_fwd\n",
        "    print(f\"Oligo 5'-Fwd: {oligo_5_prime_Fwd} {enzyme_fwd} --> primer sequence'\")\n",
        "  else:\n",
        "    print(f'Error: {enzyme_fwd} not found in RE database')\n",
        "else:\n",
        "  oligo_5_prime_Fwd2 = oligo_5_prime_Fwd\n",
        "  print(f\"Fwd primer 5'- Oligosnucleotides were not added.\")\n",
        "#@markdown\n",
        "\n",
        "oligo_5_prime_rv = 'AA' #@param {type:\"string\"}\n",
        "if not validate_oligo(oligo_5_prime_rv) and oligo_5_prime_rv != '':\n",
        "  print(f'Error: {oligo_5_prime_rv} is not a valid oligo sequence.')\n",
        "  oligo_5_prime_rv = ''\n",
        "\n",
        "enzyme_rv = 'EcoRI'  #@param {type:'string'}\n",
        "if not enzyme_fwd == '':\n",
        "  if enzyme_rv in rb_supp:\n",
        "    enzime_site_r = rb_supp.get(enzyme_rv).site\n",
        "    oligo_5_prime_rv2 = oligo_5_prime_rv + enzime_site_r\n",
        "    print(f\"Oligo 5'-Rv : {oligo_5_prime_rv} {enzime_site_r } --> primer sequence'\")\n",
        "  elif validate_oligo(enzyme_rv):\n",
        "    oligo_5_prime_rv2 = oligo_5_prime_rv + enzyme_rv\n",
        "    print(f\"Oligo 5'-Rv : {oligo_5_prime_rv} {enzyme_rv} --> primer sequence'\")\n",
        "  else:\n",
        "    print(f'Error: {enzyme_rv} not found in RE database')\n",
        "else:\n",
        "  oligo_5_prime_rv2 = oligo_5_prime_rv\n",
        "  print(f\"Rv primer 5'- Oligosnucleotides were not added.\")\n",
        "\n",
        "lfws2 =len(oligo_5_prime_Fwd2)\n",
        "lrs2 = len(oligo_5_prime_rv2)\n",
        "avg_oli=(lfws2+lrs2)/2"
      ],
      "metadata": {
        "cellView": "form",
        "id": "qG8FWrA-CN65"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#***Run primer3 for primer design with default options***"
      ],
      "metadata": {
        "id": "h7-kpbT6plSd"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title ##Run Primer3 program\n",
        "#@markdown ###For Primer3 options, see the [Primer 3 manual](https://primer3.org/manual.html).\n",
        "#@markdown <br>\n",
        "\n",
        "#**Input file options:**<br>SEQUENCE_ID=<br>SEQUENCE_TEMPLATE=<br>PRIMER_TASK=<br>PRIMER_EXPLAIN_FLAG=<<br>SEQUENCE_OVERHANG_LEFT=<br>SEQUENCE_OVERHANG_RIGHT=<br>PRIMER_PRODUCT_SIZE_RANGE=<br>SEQUENCE_FORCE_RIGHT_START=\n",
        "\n",
        "#@markdown ### **Primer length options** (without 5'- oligonucleotides)\n",
        "PRIMER_OPT_SIZE=18 #@param {type:\"integer\"}\n",
        "PRIMER_MIN_SIZE=15 #@param {type:\"integer\"}\n",
        "PRIMER_MAX_SIZE=25 #@param {type:\"integer\"}\n",
        "\n",
        "#@markdown ###<br> **Tm options**\n",
        "PRIMER_OPT_TM=60 #@param {type:\"integer\"}\n",
        "PRIMER_MAX_TM=63 #@param {type:\"integer\"}\n",
        "PRIMER_MIN_TM=57 #@param {type:\"integer\"}\n",
        "PRIMER_PAIR_MAX_DIFF_TM = 3 #@param {type:\"integer\"}\n",
        "\n",
        "#@markdown #### <br> To add custom options please upload a txt file with custom Primer3 configurations\n",
        "Custom_Options = False #@param {type:\"boolean\"}\n",
        "Custom_Options_file = '' #@param {type:'string'}\n",
        "\n",
        "if Custom_Options:\n",
        "  print('A custom configuration file will be used.')\n",
        "  if os.path.exists(Custom_Options_file):\n",
        "    configuration_file = Custom_Options_file\n",
        "  else:\n",
        "    print('Please upload your configuration file.')\n",
        "    from google.colab import files\n",
        "    uploaded = files.upload()\n",
        "    if len(uploaded) > 0:\n",
        "      configuration_file = uploaded.keys()[0]\n",
        "else:\n",
        "  configuration_file = 'input.txt'\n",
        "  with open(configuration_file,'w') as f:\n",
        "    f.write(f'''SEQUENCE_ID={cIndex}-{chromosomes[cIndex][0]}|{start}-{end}\n",
        "SEQUENCE_TEMPLATE={targetSeq.seq}\n",
        "PRIMER_TASK=generic\n",
        "PRIMER_EXPLAIN_FLAG=1\n",
        "PRIMER_PICK_ANYWAY=1\n",
        "PRIMER_OPT_SIZE={int(PRIMER_OPT_SIZE)}\n",
        "PRIMER_MIN_SIZE={int(PRIMER_MIN_SIZE)}\n",
        "PRIMER_MAX_SIZE={int(PRIMER_MAX_SIZE)}\n",
        "PRIMER_OPT_TM={PRIMER_OPT_TM}\n",
        "PRIMER_MAX_TM={PRIMER_MAX_TM}\n",
        "PRIMER_MIN_TM={PRIMER_MIN_TM}\n",
        "PRIMER_PAIR_MAX_DIFF_TM={PRIMER_PAIR_MAX_DIFF_TM}\\n''')\n",
        "    #SEQUENCE_OVERHANG_LEFT / SEQUENCE_OVERHANG_RIGHT\n",
        "    if len(oligo_5_prime_Fwd2) > 0:\n",
        "      f.write(f'SEQUENCE_OVERHANG_LEFT={oligo_5_prime_Fwd2}\\n')\n",
        "    if len(oligo_5_prime_rv2) > 0:\n",
        "      f.write(f'SEQUENCE_OVERHANG_RIGHT={oligo_5_prime_rv2}\\n')\n",
        "    if prangemin != prangemax:\n",
        "      f.write(f'''PRIMER_PRODUCT_SIZE_RANGE={prangemin}-{prangemax}\\n''')\n",
        "    if FORCE_LEFT_START:\n",
        "      f.write(f'''SEQUENCE_FORCE_LEFT_START={inStart}\\n''')\n",
        "    if FORCE_RIGHT_START:\n",
        "      f.write(f'''SEQUENCE_FORCE_RIGHT_START={inEnd}\\n''')\n",
        "    if Custom_Options:\n",
        "      with open(coptions,'r') as cop:\n",
        "        for l in cop:\n",
        "          f.write(l)\n",
        "    f.write(f'=')\n",
        "\n",
        "import os\n",
        "output = 'primers.txt'\n",
        "if os.path.exists(output):\n",
        "  os.remove(output)\n",
        "\n",
        "try:\n",
        "  os.system(f'primer3_core < {configuration_file} > primers.txt')\n",
        "  print(f'The primers were successfully designed: primers.txt')\n",
        "except:\n",
        "  print('Error designing primers')"
      ],
      "metadata": {
        "id": "I3J7OIdaJcrf",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title ##Load the primers and generate the primer table.\n",
        "#@markdown Reads the primer3 results and extracts the primer information\n",
        "import pandas as pd\n",
        "\n",
        "with open('primers.txt','r') as f:\n",
        "  primers = f.read()\n",
        "\n",
        "for l in primers.splitlines():\n",
        "  if l.startswith('PRIMER_LEFT_NUM_RETURNED='):\n",
        "    nLP = int(l.split('=')[1])\n",
        "  elif l.startswith('PRIMER_RIGHT_NUM_RETURNED='):\n",
        "    nRP = int(l.split('=')[1])\n",
        "\n",
        "if (nLP > 0) and (nRP > 0):\n",
        "  nprimers = min(nLP,nRP)\n",
        "  #read primers\n",
        "  primersDict = {}\n",
        "  for i in range(nprimers):\n",
        "    primerInfo = {}\n",
        "    for l in primers.splitlines():\n",
        "      if  l.startswith(f'PRIMER_PAIR_{i}_PENALTY'):\n",
        "        primerInfo['PairPenalty'] = l.strip().split('=')[1]\n",
        "      elif  l.startswith(f'PRIMER_LEFT_{i}_PENALTY'):\n",
        "        primerInfo['fwdPenalty'] = l.strip().split('=')[1]\n",
        "      elif  l.startswith(f'PRIMER_RIGHT_{i}_PENALTY'):\n",
        "        primerInfo['rvPenalty'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_LEFT_{i}_SEQUENCE'):\n",
        "        primerInfo['fwdSeq'] = l.strip().split('=')[1]\n",
        "        primerInfo['fwdLen'] = len(l.strip().split('=')[1])\n",
        "      elif l.startswith(f'PRIMER_RIGHT_{i}_SEQUENCE'):\n",
        "        primerInfo['rvSeq'] = l.strip().split('=')[1]\n",
        "        primerInfo['rvLen'] = len(l.strip().split('=')[1])\n",
        "      elif l.startswith(f'PRIMER_LEFT_{i}_TM'):\n",
        "        primerInfo['fwdTM'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_RIGHT_{i}_TM'):\n",
        "        primerInfo['rvTM'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_LEFT_{i}_GC_PERCENT'):\n",
        "        primerInfo['fwdGC'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_RIGHT_{i}_GC_PERCENT'):\n",
        "        primerInfo['rvGC'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_LEFT_{i}_PROBLEMS'):\n",
        "        primerInfo['fwdProblems'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_RIGHT_{i}_PROBLEMS'):\n",
        "        primerInfo['rvProblems'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_LEFT_{i}_SELF_ANY_TH'):\n",
        "        primerInfo['fwdSelfAny'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_RIGHT_{i}_SELF_ANY_TH'):\n",
        "        primerInfo['rvSelfAny'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_LEFT_{i}_SELF_END_TH'):\n",
        "        primerInfo['fwdSelfEnd'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_RIGHT_{i}_SELF_END_TH'):\n",
        "        primerInfo['rvSelfEnd'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_LEFT_{i}_HAIRPIN_TH'):\n",
        "        primerInfo['fwdHairpin'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_RIGHT_{i}_HAIRPIN_TH'):\n",
        "        primerInfo['rvHairpin'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_LEFT_{i}_END_STABILITY'):\n",
        "        primerInfo['fwdEndStability'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_RIGHT_{i}_END_STABILITY'):\n",
        "        primerInfo['rvEndStability'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_PAIR_{i}_COMPL_ANY_TH'):\n",
        "        primerInfo['complAny'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_PAIR_{i}_COMPL_END_TH'):\n",
        "        primerInfo['complEnd'] = l.strip().split('=')[1]\n",
        "      elif l.startswith(f'PRIMER_PAIR_{i}_PRODUCT_SIZE'):\n",
        "        primerInfo['productSize'] = l.strip().split('=')[1]\n",
        "      elif  l.startswith(f'PRIMER_PAIR_{i}_PRODUCT_TM'):\n",
        "        primerInfo['productTM'] = l.strip().split('=')[1]\n",
        "    primersDict[i] = primerInfo\n",
        "\n",
        "primerTable = pd.DataFrame.from_dict(primersDict, orient='index', columns = ['PairPenalty',\n",
        "'fwdPenalty','rvPenalty','fwdSeq','fwdLen','rvSeq','rvLen','fwdTM','rvTM','fwdGC','rvGC','fwdProblems','rvProblems',\n",
        "'fwdSelfAny','rvSelfAny','fwdSelfEnd','rvSelfEnd','fwdHairpin','rvHairpin','fwdEndStability',\n",
        "'rvEndStability','complAny','complEnd','productSize','productTM'])\n",
        "# (names=['name','TM','Secuencia','Comentarios'])\n",
        "primerTable.to_csv('primerTable.csv')\n",
        "\n",
        "if len(primerTable) == 0:\n",
        "  print('Primer3 could not find any primers with the provided options. Try relaxing Tm settings.')\n",
        "primerTable"
      ],
      "metadata": {
        "id": "jeiqDcLDLkx9",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#***Blastn Virtual PCR***."
      ],
      "metadata": {
        "id": "aMli4XKxWMUn"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title #Search for the primer sequences in the target genome\n",
        "#@markdown ###This block searches for all the primer sets in the target genome.\n",
        "\n",
        "with open('virtualPCR.csv','w') as f:\n",
        "  f.write('SSeqID,SStart,SEnd,Fwd,FwdLen,#FwdFound,#FwdFoundWithMismatch,AlnLength,FwdSeq,SSeq,PIdent,Mismatch,Rv,RvLen,#RvFound,#RvFoundWithMismatch,AlnLength,RvSeq,SSeq,PIdent,Mismatch,ProductSize\\n')\n",
        "#@markdown ###<br> **Maximum difference in primer vs alignment length**\n",
        "max_diff_aln_len = 3 #@param {type:\"integer\"}\n",
        "#@markdown ####<br> Should 5'-oligos be removed from primers for virtual PCR?\n",
        "remove_oligos = True #@param {type:'boolean'}\n",
        "\n",
        "#primerTable\n",
        "verify_first_primer_Set_only = True  #@param {type:\"boolean\"}\n",
        "\n",
        "if verify_first_primer_Set_only:\n",
        "  r = primerTable.iloc[0]\n",
        "  i = 0\n",
        "  q = 'primer_pair.tmp'\n",
        "  with open(q,'w') as f:\n",
        "    f.write(f\">{i}_Fwd\\n{str(r['fwdSeq'])}\\n>{i}_rv\\n{str(r['rvSeq'])}\")\n",
        "  print(f\"Procesando primers {i}_Fwd:{str(r['fwdSeq'])} | {i}_rv:{str(r['rvSeq'])}\")\n",
        "  numProducts = primer_blast(q,oligo_5_prime_Fwd2,oligo_5_prime_rv2,gfile,i,max_diff_aln_len,remove_oligos)\n",
        "else:\n",
        "  for i,r in primerTable.iterrows():\n",
        "    q = 'primer_pair.tmp'\n",
        "    with open(q,'w') as f:\n",
        "      f.write(f\">{i}_Fwd\\n{str(r['fwdSeq'])}\\n>{i}_rv\\n{str(r['rvSeq'])}\")\n",
        "    print(f\"Procesando primers {i}_Fwd:{str(r['fwdSeq'])} | {i}_rv:{str(r['rvSeq'])}\")\n",
        "    numProducts = primer_blast(q,oligo_5_prime_Fwd2,oligo_5_prime_rv2,gfile,i,max_diff_aln_len,remove_oligos)\n",
        "\n",
        "print(f'The results are in the virtualPCR.csv file.')\n",
        "table = pd.read_csv('virtualPCR.csv')\n",
        "print('The complete filtered blast table is in the FilteredPrimerHits_[primer pair].csv file.')\n",
        "table"
      ],
      "metadata": {
        "id": "bRI_4bz6OsBm",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title ###Search in the target genome custom designed primers\n",
        "#@markdown ###VirtualPCR with custom primers\n",
        "#@markdown [Requires a genome file to be downloaded or a local genome file]\n",
        "\n",
        "#@markdown #### **Input the custom primers**\n",
        "import pandas as pd\n",
        "if os.path.exists('virtualPCR.csv'):\n",
        "  os.remove('virtualPCR.csv')\n",
        "\n",
        "with open('virtualPCR.csv','w') as f:\n",
        "  f.write('SSeqID,SStart,SEnd,Fwd,FwdLen,#FwdFound,#FwdFoundWithMismatch,AlnLength,FwdSeq,SSeq,PIdent,Mismatch,Rv,RvLen,#RvFound,#RvFoundWithMismatch,AlnLength,RvSeq,SSeq,PIdent,Mismatch,ProductSize\\n')\n",
        "\n",
        "oligo_5_prime_Fwd_custom = '' #@param {type:\"string\"}\n",
        "specific_sequence_Fwd = '' #@param {type:\"string\"}\n",
        "\n",
        "oligo_5_prime_rv_custom = '' #@param {type:\"string\"}\n",
        "specific_sequence_rv = '' #@param {type:\"string\"}\n",
        "\n",
        "#@markdown ####<br> Should 5'-oligos be removed from primers for virtual PCR?\n",
        "remove_oligos = True #@param {type:'boolean'}\n",
        "\n",
        "#@markdown ###<br>Maximum difference in primer vs alignment length\n",
        "max_diff_aln_len = 3 #@param {type:\"integer\"}\n",
        "\n",
        "q = 'primer_pair.tmp'\n",
        "with open(q,'w') as f:\n",
        "  f.write(f\">Custom_Fwd\\n{oligo_5_prime_Fwd_custom}{specific_sequence_Fwd}\\n>Custom_Rv\\n{oligo_5_prime_rv_custom}{specific_sequence_rv}\")\n",
        "\n",
        "numProducts = primer_blast(q,oligo_5_prime_Fwd_custom,oligo_5_prime_rv_custom,gfile,'custom',max_diff_aln_len,remove_oligos)\n",
        "table = pd.read_csv('virtualPCR.csv')\n",
        "table\n"
      ],
      "metadata": {
        "id": "KtOuWizyXb24",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}